### PR TITLE
DOC: Replace Sphinx search with Algolia

### LINKED ISSUE #17037: DOC: Add a better search engine
We have an opportunity to get better doc search on our site,  with the new numpy.org thrown in as well.

It's the search engine pytorch.org uses. If you do a search there and click the `Algolia` logo at the bottom right, [you'll see the deal](https://www.algolia.com/docsearch): Algolia provides pytorch.org with search for free.

Algolia is looking to work with sites like ours -- sites hosting open-source docs. 

Better site search would help people get the right content for the right NumPy version, rather than spinning the wheel with Google.

Does anyone object if I get started with this? They'll crawl our site and get back to us with an API key. 

I'll integrate it; it's been done with Sphinx before.

### PR DESCRIPTION
**Pull request by bjnath**
'Quick search' removed from sidebars; all pages now have a searchbar at the top. Sphinx search remains accessible in a dedicated page.

This is the search used in pytorch.org (among [many others](https://docsearch.algolia.com/)).

I'll post the CI artifact to demonstrate the look and feel. However,  WIP -- always says "no results found."  The PR gets a live version up for Algolia to take a look at.

Closes gh-17037.

### CODE REVIEW

**Comment by mattip** [comment]
xref issue gh-17037

I think we should only change the search engine in this PR and not touch the formatting. One thing at a time.

**Comment by bjnath** [comment]
There's no formatting change. This is based on master. New logo went in with #16988.

**Comment by bjnath** [comment]
Artifact: https://15086-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

File: doc/source/_static/numpy.css [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+div.spc-navbar .nav-pills > li > a {
+  background-color: #4d77cf; /* Han Blue */
+}
+
+/* Eliminate the Quick Search boxes in the sidebar;
+the search results get clipped by the sidebar margin */
+#searchbox {
+  visibility: hidden;
+}
```
Comment (mattip) [inline doc/source/_static/numpy.css:32]: This seems like a formatting change, no?

File: doc/source/_templates/layout.html [hunk 2]
```diff
@@ -1,35 +1,22 @@
...
-    box-shadow: none;
-  }
-  div.top-scipy-org-logo-header img {
-    height: 100px;
-    padding-left: 27px;
-  }
-  div.spc-navbar .nav-pills > li > a {
-    background-color: #4d77cf; /* Han Blue */
-  }
-</style>
```
Comment (eric-wieser) [inline doc/source/_templates/layout.html:23]: Can you move this code around in a standalone PR?

File: doc/source/_templates/layout.html [hunk 3]
```diff
@@ -1,35 +1,22 @@
...
 {{ super() }}
+<link rel="stylesheet" href="{{ pathto('_static/numpy.css', 1) }}" type="text/css" />
+<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
 {% endblock %}
 
 {%- block header %}
 <div class="container">
   <div class="top-scipy-org-logo-header">
-    <a href="{{ pathto('index') }}">
```
Comment (eric-wieser) [inline doc/source/_templates/layout.html:30]: This looks like an unrelated change too

File: doc/source/_templates/layout.html [hunk 4]
```diff
@@ -52,3 +39,17 @@
...
 {{ super() }}
 {% endblock %}
+
+{%- block footer %}
+{{ super() }}
+<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
+<script type="text/javascript"> docsearch({
+apiKey: 'c8367ae0f87be4fd19cb01bae97ef4f2',
+indexName: 'numpy',
+inputSelector: 'form input[type=text]',
```
Comment (eric-wieser) [inline doc/source/_templates/layout.html:49]: This seems dangerously vague - can we use a more specific filter?
```suggestion
inputSelector: '#algolia-searchbar form input[type=text]',
```

File: doc/source/_templates/layout.html (hunk 2 above)
Comment (bjnath) [reply doc/source/_templates/layout.html:23]: This code is already in #16988. I've moved it out of the header. This is not two PRs in one.

File: doc/source/_templates/layout.html (hunk 3 above)
Comment (bjnath) [reply doc/source/_templates/layout.html:30]: That is true.

File: doc/source/_templates/layout.html (hunk 4 above)
Comment (bjnath) [reply doc/source/_templates/layout.html:49]: Agreed.

File: doc/source/_templates/layout.html (hunk 2 above)
Comment (eric-wieser) [reply doc/source/_templates/layout.html:23]: I'm suggesting making a PR that does nothing but move this code. It makes your other PRs conflict less, and makes all the PRs easier to review.

File: doc/source/_static/numpy.css (hunk 1 above)
Comment (bjnath) [reply doc/source/_static/numpy.css:32]: This is the Sphinx "Quick search" searchbar that's being hidden, as noted above:

> 'Quick search' removed from sidebars

This is done so pages won't have two unrelated searchbars in different locations.

**Comment by bashtage** [comment]
This implementation is different from how pytorch works.  In pytorch, the equivalent search box on this page:

https://15086-908607-gh.circle-artifacts.com/0/doc/build/html/reference/index.html

uses standard sphinx search.

File: doc/source/_templates/layout.html (hunk 2 above)
Comment (bjnath) [reply doc/source/_templates/layout.html:23]: This makes sense and I'll bear it in mind for future PRs.

**Comment by bjnath** [comment]
I've changed the PR title and comments to clarify that Sphinx search is still there and still usable.

**Comment by bashtage** [comment]
I don't see how this is opt-in.  I would think that opt-in would require checking an environmental variable like `NUMPY_USE_ALGOLIA` and this code if 1 (or true, or equivalent).

**Comment by rgommers** [comment]
> uses standard sphinx search.
> 
> Mirroring pytorch would be placing this search on the numpy.org website, not in the reference documentation.

This is a good point. It does make sense to put this on `numpy.org` only. If we do that, I'd like to unify the main docs and NEPs builds again though. Separate deployment of http://www.numpy.org/neps/ never made much sense to me, and having the NEPs unsearchable from the docs is quite annoying.

**Comment by bjnath** [comment]
@rgommers 
> It does make sense to put this on numpy.org only.

I think users will be baffled by having search engines change underneath them. They have no reason to expect different results on different pages.

**Comment by bjnath** [comment]
> opt-in would require checking an environmental variable like NUMPY_USE_ALGOLIA 

Without yet considering the exact mechanism, does everyone see configurability an acceptable solution?

**Comment by mattip** [comment]
My concerns about adding Javascript that sends information to a commercial entity still remain

**Comment by rgommers** [comment]
> Without yet considering the exact mechanism, does everyone see configurability as an acceptable solution?

No, that doesn't make much sense to me. 

> My concerns about adding Javascript that sends information to a commercial entity still remain

If there's a better alternative I'd love to hear it, but my prio 1 is to actually _have_ a search box that covers all our content.

**Comment by rgommers** [comment]
We also have Google Analytics by the way, that sends data to a commercial entity ....

**Comment by bjnath** [comment]
> that sends information to a commercial entity

Rather than battling this a PR at a time, with possibly inconsistent outcomes, we should have a policy in place. 

Submitted issue #17066

**Comment by mattip** [comment]
> We also have Google Analytics by the way, that sends data to a commercial entity ....
> And Cloudflare, that keeps stats on all our visitors ....

We should definitely reconsider Google Analytics. More lightweight and transparent alternatives exist. I don't think anyone actually considered the implications when adding it, and a quick search will show that many people have issues with it.

I don't know much about data collection via CloudFare, but that is a different story: they actually provide a useful caching service.

**Comment by bjnath** [comment]
The search engine is working at https://15101-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

Even if you think these guys are devils and would never put Algolia on our site, you it to yourself to have a look.

Content from NEPs and numpy.org are included.

**Comment by bashtage** [comment]
How do you get a paginate search if you need to look beyond the top 5?  I couldn't get it to work on mobile.

Also, it is duplicating the results. For example, try entering the start of concatenate.

**Comment by bjnath** [comment]
@bashtage
> Also, it is duplicating the results. 

One is from devdocs, one is from stable.  Making the choice of dev/stable user-selectable so you see only one or the other.

**Comment by bjnath** [comment]
> How do you get a paginate search if you need to look beyond the top 5?

The search on their site has scrollbars; I've asked why ours doesn't.

**Comment by charris** [comment]
Ping all.

**Comment by mattip** [comment]
Let's close this in the hopes that I will finish gh-17265 soonish.
